File: Objective-C/macOS/PurchasesDetails.m
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
A PrimaryViewController subclass that is used to display purchases or restored ones. Provides details about a selected purchase. The purchase con-
 tains the product identifier, transaction identifier, and transaction date for a regular purchase; the content identifier, content version, and
 content length for a hosted product; and the original transaction's identifier and date for a restored product.
*/

@import StoreKit;
#import "StoreManager.h"
#import "PurchasesDetails.h"
#import "NSDateFormatter+DateFormatter.h"
#import "SKDownload+SKDownloadAdditions.h"

@interface PurchasesDetails ()
@property (weak) IBOutlet NSStackView *stackView;
@property (weak) IBOutlet NSTextField *productID;
@property (weak) IBOutlet NSTextField *transactionID;
@property (weak) IBOutlet NSTextField *transactionDate;

@property (weak) IBOutlet NSBox *download;
@property (weak) IBOutlet NSTextField *downloadID;
@property (weak) IBOutlet NSTextField *contentVersion;
@property (weak) IBOutlet NSTextField *contentLength;

@property (weak) IBOutlet NSBox *originalTransaction;
@property (weak) IBOutlet NSTextField *originalTransactionID;
@property (weak) IBOutlet NSTextField *originalTransactionDate;
@end

@implementation PurchasesDetails
#pragma mark - View Life Cycle

- (void)viewDidAppear {
	[super viewDidAppear];

	[self hideBox:self.download forStatus:YES];
	[self hideBox:self.originalTransaction forStatus:YES];
	[self reloadTableAndSelectFirstRowIfNecessary];
}

#pragma mark - Update UI

/// Hides an NSBox object if the specified status is true and shows it, otherwise.
-(void)hideBox:(NSBox *)box forStatus:(BOOL)status {
	if (self.viewLoaded) {
		box.hidden = status;
		[self.view layoutSubtreeIfNeeded];
	}
}

/// Reloads the UI with new products, invalid identifiers, or purchases.
-(void)reloadUIWithData:(NSMutableArray *)data {
	self.data = data;
	[self.tableView reloadData];
	[self reloadTableAndSelectFirstRowIfNecessary];
}

/// Reloads the table view and programmatically selects a purchase.
-(void)reloadTableAndSelectFirstRowIfNecessary {
    // Select the first purchase and display its information if no row are currently selected and display the current selection, otherwise.
	NSIndexSet *selection = (self.tableView.selectedRowIndexes.count == 0) ? [NSIndexSet indexSetWithIndex:0] : self.tableView.selectedRowIndexes;

	[self.tableView reloadData];
	[self.tableView selectRowIndexes:selection byExtendingSelection:NO];
}

#pragma mark - NSTableViewDelegate

- (NSView *)tableView:(NSTableView *)tableView viewForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row {
	SKPaymentTransaction *transaction = (SKPaymentTransaction *)self.data[row];
	NSTableCellView *cell = [tableView makeViewWithIdentifier:tableColumn.identifier owner:self];

	if (cell != nil) {
		// Display the product's title associated with the payment's product identifier.
		cell.textField.stringValue = [[StoreManager sharedInstance] titleMatchingPaymentTransaction:transaction];
		return cell;
	}
	return nil;
}

/// Displays information about the selected purchase or restored one.
- (void)tableViewSelectionDidChange:(NSNotification *)notification {
	NSInteger selectedRow = (self.tableView).selectedRow;
	NSDateFormatter *dateFormatter = [NSDateFormatter longStyle];

	[self hideBox:self.download forStatus:YES];
	[self hideBox:self.originalTransaction forStatus:YES];

	if (selectedRow >= 0 && (self.data.count > 0)) {
		SKPaymentTransaction *transaction = (SKPaymentTransaction *)self.data[selectedRow];

		self.productID.stringValue = transaction.payment.productIdentifier;
		self.transactionID.stringValue = transaction.transactionIdentifier;
		self.transactionDate.stringValue = [dateFormatter stringFromDate:transaction.transactionDate];

		NSArray *allDownloads = transaction.downloads;

		// Display download information if they are any.
		if (allDownloads.count > 0) {
			[self hideBox:self.download forStatus:NO];

            // We are only showing the first download.
			SKDownload *firstDownload = allDownloads[0];
			self.downloadID.stringValue = firstDownload.contentIdentifier;
			self.contentVersion.stringValue = firstDownload.contentVersion;
			self.contentLength.stringValue = firstDownload.downloadContentSize;
		}

		// Display restored transactions if they exist.
		if (transaction.originalTransaction != nil) {
			[self hideBox:self.originalTransaction forStatus:NO];

			self.originalTransactionID.stringValue = transaction.originalTransaction.transactionIdentifier;
			self.originalTransactionDate.stringValue = [dateFormatter stringFromDate:transaction.originalTransaction.transactionDate];
		}
	}
}

@end

File: Swift/macOS/PurchasesDetails.swift
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
A PrimaryViewController subclass that is used to display purchases or restored ones. Provides details about a selected purchase. The purchase con-
tains the product identifier, transaction identifier, and transaction date for a regular purchase; the content identifier, content version, and
content length for a hosted product; and the original transaction's identifier and date for a restored product.
*/

import Cocoa
import StoreKit

class PurchasesDetails: PrimaryViewController {
	// MARK: - Properties

	@IBOutlet weak fileprivate var stackView: NSStackView!
	@IBOutlet weak fileprivate var productID: NSTextField!
	@IBOutlet weak fileprivate var transactionID: NSTextField!
	@IBOutlet weak fileprivate var transactionDate: NSTextField!

	@IBOutlet weak fileprivate var download: NSBox!
	@IBOutlet weak fileprivate var downloadID: NSTextField!
	@IBOutlet weak fileprivate var contentVersion: NSTextField!
	@IBOutlet weak fileprivate var contentLength: NSTextField!

	@IBOutlet weak fileprivate var originalTransaction: NSBox!
	@IBOutlet weak fileprivate var originalTransactionID: NSTextField!
	@IBOutlet weak fileprivate var originalTransactionDate: NSTextField!

	// MARK: - View Life Cycle

	override func viewDidAppear() {
		super.viewDidAppear()
		download.hide()
		originalTransaction.hide()
		reloadTableAndSelectFirstRowIfNecessary()
	}

	// MARK: - Update UI

	/// Refreshes the UI with new payment transactions.
	func reload(with transactions: [SKPaymentTransaction]) {
		data = transactions
		self.tableView.reloadData()
		reloadTableAndSelectFirstRowIfNecessary()
	}

	/// Reloads the table view and programmatically selects a purchase.
	fileprivate func reloadTableAndSelectFirstRowIfNecessary() {
		// Select the first purchase and display its information if no row are currently selected and display the current selection, otherwise.
		let selection = (tableView.selectedRowIndexes.isEmpty) ? IndexSet(integer: 0) : tableView.selectedRowIndexes

		tableView.reloadData()
		tableView.selectRowIndexes(selection, byExtendingSelection: false)
	}

	// MARK: - NSTableViewDelegate

	func tableView(_ tableView: NSTableView, viewFor tableColumn: NSTableColumn?, row: Int) -> NSView? {
		guard let tableColumn = tableColumn, let cell = tableView.makeView(withIdentifier: tableColumn.identifier, owner: nil) as? NSTableCellView,
			let transaction = data[row] as? SKPaymentTransaction else { return nil }

		// Display the product's title associated with the payment's product identifier.
		cell.textField?.stringValue = StoreManager.shared.title(matchingPaymentTransaction: transaction)
		return cell
	}

	/// Displays information about the selected purchase or restored one.
	func tableViewSelectionDidChange(_ notification: Notification) {
		let selectedRow = tableView.selectedRow
		guard (selectedRow >= 0) && (!data.isEmpty) else { return }

		download.hide()
		originalTransaction.hide()

		guard let transaction = data[selectedRow] as? SKPaymentTransaction else { return }

		productID.stringValue = transaction.payment.productIdentifier
		transactionID.stringValue = transaction.transactionIdentifier!
		transactionDate.stringValue = DateFormatter.long(transaction.transactionDate!)

		let allDownloads = transaction.downloads
		// Display download information if they are any.
		if !allDownloads.isEmpty {
			download.show()

			// We are only showing the first download.
			if let firstDownload = allDownloads.first {
				downloadID.stringValue = firstDownload.contentIdentifier
				contentVersion.stringValue = firstDownload.contentVersion
				contentLength.stringValue = firstDownload.downloadContentSize
			}
		}
		// Display restored transactions if they exist.
		guard let transactionIdentifier = transaction.original?.transactionIdentifier, let transactionDate = transaction.original?.transactionDate
			else { return }

		originalTransaction.show()
		originalTransactionID.stringValue = transactionIdentifier
		originalTransactionDate.stringValue = DateFormatter.long(transactionDate)
	}
}
